cmd/metal-api/internal/scaler/poolscaler.go

```diff
@@ -0,0 +1,178 @@
+package scaler
+
+import (
+	"math"
+	"math/rand"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/metal-stack/metal-api/cmd/metal-api/internal/metal"
+	"go.uber.org/zap"
+)
+
+type (
+	PoolScaler struct {
+		log     *zap.SugaredLogger
+		manager MachineManager
+	}
+	PoolScalerConfig struct {
+		Log     *zap.SugaredLogger
+		Manager MachineManager
+	}
+	MachineManager interface {
+		AllMachines() (metal.Machines, error)
+		WaitingMachines() (metal.Machines, error)
+		ShutdownMachines() (metal.Machines, error)
+
+		Shutdown(m *metal.Machine) error
+		PowerOn(m *metal.Machine) error
+	}
+)
+
+func NewPoolScaler(log *zap.SugaredLogger, manager MachineManager) *PoolScaler {
+	return &PoolScaler{
+		log:     log,
+		manager: manager,
+	}
+}
+
+// AdjustNumberOfWaitingMachines compares the number of waiting machines to the required pool size of the partition and shuts down or powers on machines accordingly
+func (p *PoolScaler) AdjustNumberOfWaitingMachines(partition *metal.Partition) error {
+	waitingMachines, err := p.manager.WaitingMachines()
+	if err != nil {
+		return err
+	}
+
+	poolSizeExcess := 0
+	poolSizeExcess = p.calculatePoolSizeExcess(len(waitingMachines), partition.WaitingPoolMinSize, partition.WaitingPoolMaxSize)
+
+	if poolSizeExcess == 0 {
+		return nil
+	}
+
+	if poolSizeExcess > 0 {
+		p.log.Infow("shut down spare machines", "number", poolSizeExcess, "partition", p)
+		err := p.shutdownMachines(waitingMachines, poolSizeExcess)
+		if err != nil {
+			return err
+		}
+	} else {
+		shutdownMachines, err := p.manager.ShutdownMachines()
+		if err != nil {
+			return err
+		}
+
+		poolSizeExcess = int(math.Abs(float64(poolSizeExcess)))
+		if len(shutdownMachines) < poolSizeExcess {
+			p.log.Infow("not enough machines to meet waiting pool size; power on all remaining", "number", len(shutdownMachines), "partition", p)
+			err = p.powerOnMachines(shutdownMachines, len(shutdownMachines))
+			if err != nil {
+				return err
+			}
+			return nil
+		}
+
+		p.log.Infow("power on missing machines", "number", poolSizeExcess, "partition", p)
+		err = p.powerOnMachines(shutdownMachines, poolSizeExcess)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// calculatePoolSizeExcess checks if there are less waiting machines than minRequired or more than maxRequired
+// if yes, it returns the difference between the actual amount of waiting machines and the average of minRequired and maxRequired
+// if no, it returns 0
+func (p *PoolScaler) calculatePoolSizeExcess(actual int, minRequired, maxRequired string) int {
+	if strings.Contains(minRequired, "%") != strings.Contains(maxRequired, "%") {
```

Comment:
`strings.HasSuffix` could be a bit more appropriate.

Feels like handling the "percent or int" property makes the function a little harder to read. Maybe it would be a nice idea to factor it out into a small helper type called `ScalerRange` in the `metal/partition.go` with small functions like `IsPercentage` or even just `GetMin` and `GetMax` (figuring out / calculating the value by itself)? You can then also put your validations in there.